File: web/src/lib/precision.ts
/**
 * Precision classification for confidence intervals.
 *
 * Mirrors ``backend/chart/risk/precision.py`` — keep the thresholds in
 * lockstep with that module so the badge shown by the frontend matches
 * the one the backend computes for the horizon cards.
 *
 * We classify precision by the CI ratio ``high / low`` rather than the
 * raw width because ratio is scale-invariant: the same thresholds work
 * for raw odds ratios (``1.03``) and milli-encoded values (``1030``).
 *
 * Thresholds agreed with the health team (Alessandro / Lucia, 2026-08):
 *   - HIGH — CI ratio ≤ 2.5 (no indication of substantial imprecision)
 *   - MODERATE — 2.5 < CI ratio ≤ 5 (potential imprecision)
 *   - LOW — CI ratio > 5 (imprecise / wide confidence interval)
 */

export type PrecisionLevel = "high" | "moderate" | "low";

export const HIGH_CI_RATIO_MAX = 2.5;
export const MODERATE_CI_RATIO_MAX = 5;

export function ciRatio(low: number, high: number): number {
  if (!Number.isFinite(low) || !Number.isFinite(high) || low <= 0) {
    return Number.POSITIVE_INFINITY;
  }
  return Math.max(low, high) / low;
}

export function precisionFromCi(low: number, high: number): PrecisionLevel {
  const ratio = ciRatio(low, high);
  if (ratio <= HIGH_CI_RATIO_MAX) return "high";
  if (ratio <= MODERATE_CI_RATIO_MAX) return "moderate";
  return "low";
}

const DEFAULT_LABEL: Record<PrecisionLevel, string> = {
  high: "High",
  moderate: "Moderate",
  low: "Low",
};

export function precisionLabel(level: PrecisionLevel): string {
  return DEFAULT_LABEL[level];
}

File: backend/chart/risk/precision.py
"""Precision classification for confidence intervals.

The dashboard's precision badge summarises how tight a 95% CI is around
a relative-risk / odds-ratio estimate. Rather than reasoning about the
raw CI width (which is scale-dependent), we use the CI ratio
``high / low`` — this is scale-invariant so the same thresholds work
whether the caller passes raw floats (RR = 1.15) or milli-scaled ints
(1150).

Thresholds agreed with the health team (see Alessandro / Lucia,
2026-08):

* HIGH precision — CI ratio ≤ 2.5 (no indication of substantial imprecision)
* MODERATE precision — 2.5 < CI ratio ≤ 5 (potential imprecision)
* LOW precision — CI ratio > 5 (imprecise / wide confidence interval)

Keep these in sync with ``web/src/lib/precision.ts``.
"""

from __future__ import annotations

from typing import Literal


PrecisionLevel = Literal["high", "moderate", "low"]

HIGH_CI_RATIO_MAX = 2.5
MODERATE_CI_RATIO_MAX = 5.0


def ci_ratio(low: float, high: float) -> float:
    """Return the confidence-interval ratio (``high / low``).

    Defaults to ``float('inf')`` when ``low`` is non-positive, which
    forces the caller into the LOW precision bucket rather than blowing
    up on a divide-by-zero.
    """

    if low <= 0:
        return float("inf")
    return max(high, low) / low


def precision_for_ci(low: float, high: float) -> PrecisionLevel:
    """Classify a 95% CI into a precision level.

    Works for both raw estimates (``low=1.03, high=1.29``) and milli
    encodings (``low=1030, high=1290``) because the classification
    depends only on the ratio.
    """

    ratio = ci_ratio(low, high)
    if ratio <= HIGH_CI_RATIO_MAX:
        return "high"
    if ratio <= MODERATE_CI_RATIO_MAX:
        return "moderate"
    return "low"
